### PR TITLE
Plugins: Update playlist type after plugin resolution

### PR DESCRIPTION
**Pull request by enen92**
## Description
This pull-request reverts https://github.com/xbmc/xbmc/pull/16158 (originally as a fix for https://github.com/xbmc/xbmc/issues/16151) and purposes a different fix for the issue. Additionally, it might also fix other issues such as https://github.com/xbmc/xbmc/issues/17915 (untested).

## Motivation
I am trying to solve issues in python addons due to the concurrent busydialogs. Unfortunately, the problem is not easily solvable since `CPluginDirectory::GetPluginDirectory` (filesystem) at the moment includes and calls into GUI if called from the main thread. To be able to remove such dependency (first step into having a fix) we need to isolate and track which calls into `CPluginDirectory::GetPluginResult` may originate from Application (e.g. play a plugin:// path) vs those that are supposed to occur in the background (e.g. python scrappers). Unfortunately, while my previous solution solved the reported issue, it also brought a side-effect of including a new call into the aforementioned method. PlaylistPlayer can be called from anywhere (e.g. JSON) while PluginDirectory retrieval is supposed to be a busy process (guarded by a busydialog the user might want to cancel). Hence, different approaches to solve the original issue are required.

PlayList player may hold two types of playlists: `VIDEO` and `MUSIC`. Those are created when the list of items are passed to the player and by default it defaults to `MUSIC`. It only becomes `VIDEO` if any of the items in the file list can be identified as a video (see `CFileitem.IsVideo`). The problem is that plugins can hold any type of content while kodi can only infer the type after the plugin:// path has been expanded/resolved. When PlayListPlayer cannot infer the type, it calls generic `PlayMedia` in application (ugly but not the issue I'm trying to solve here) along with the playlist type. This was not taken into account for plugins:// and we could end up in situations where the current active playlist is music but we are actually playing a video.

To add to the problem, JSONRPC seems to link player id's with Playlist types (most likely the weigh of legacy). 

So, if a plugin item is played from json rpc (e.g. from yatse):

- PlaylistPlayer creates a music playlist (can't determine the plugin:// content type) so keeps the playlist as music and calls PlayMedia
- Videoplayer will be opened (since the item is actually a video)
- **GetActivePlayers** in JSON will return PlayerId=0 (music) as the current active player
- **GetPlayerProperties** will be called with PlayerId=0 (active player) which is a "music player"
- Most of the properties (subtitles, etc) will not be available in the response as they are invalid for music
This is the reason why situations such as the ones described in https://github.com/xbmc/xbmc/issues/17915 end up happening.

This PR updates the playlist type if PlayMedia is called with a playlist type, a plugin:// item was passed to the function and they types differ.

I know all the logic I described above (basically how kodi works...) is ugly and violates a lot of good practices (even the new fix is also ugly). This is not the issue I'm trying to solve here. My goal is to remove GUI from PluginDirectory (filesystem).

Not sure who to ping.

## How Has This Been Tested?
Manually, same test scenario as https://github.com/xbmc/xbmc/pull/16158

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by enen92** [comment]
@CastagnaIT can you check if this solves your issue?

File: xbmc/FileItem.cpp [hunk 1]
```diff
@@ -834,7 +834,7 @@ bool CFileItem::IsVideo() const
   //! @todo If the file is a zip file, ask the game clients if any support this
   // file before assuming it is video.
 
-  return URIUtils::HasExtension(m_strPath, CServiceBroker::GetFileExtensionProvider().GetVideoExtensions());
+  return URIUtils::HasExtension(GetDynPath(), CServiceBroker::GetFileExtensionProvider().GetVideoExtensions());
```
Comment (Rechi) [inline xbmc/FileItem.cpp:837]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/FileItem.cpp [hunk 2]
```diff
@@ -921,7 +921,7 @@ bool CFileItem::IsAudio() const
   //! @todo If the file is a zip file, ask the game clients if any support this
   // file before assuming it is audio
 
-  return URIUtils::HasExtension(m_strPath, CServiceBroker::GetFileExtensionProvider().GetMusicExtensions());
+  return URIUtils::HasExtension(GetDynPath(), CServiceBroker::GetFileExtensionProvider().GetMusicExtensions());
```
Comment (Rechi) [inline xbmc/FileItem.cpp:924]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by enen92** [comment]
Jenkins build this please.

**Comment by CastagnaIT** [comment]
hi is possible have an artifact for win64?
thank you

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by Tolriq** [comment]
Beware that Kodi support mixed playlists.
So if you play a soundclound radio and queue a youtube music video both will still be in music playlist.

Depending on when this play call is made it either fix only a small subset of the issue (playing with fresh playlist and title should be updated), or if it's called during queuing to expand added items to playlist then it will causes issues as the playlist will switch, meaning the rest of the soundcloud songs won't play.

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by enen92** [comment]
@Tolriq thanks for the note. Adding to the queue is handled by `Add()` in playlist player and doesn't do anything other than adding some unknown item to an existing list, without further resolution.  My change only affects asking the item for playback.

But now that you said that, I'll run a few more tests. 
I think the "only" thing we need is to change the playlist content to video if the original value was music after we just got a video item from plugin resolution. The other condition (the inverse) isn't probably necessary because by default kodi would have created a VIDEO playlist if it knew the item was holding a video.

I say the "only" thing we need because all this could well be refactored :( PlaylistPlayer calls into application for pretty much everything..

**Comment by Tolriq** [comment]
@enen92 if the call is made at the playtime so does the expanding of the plugin the following case will fail:
Queue soundclound, queue youtube video, queue soundcloud.

Soundcloud will play, then youtube will be detected as video, playlist switch, expand, and playlist no more contains the previous song so no back and no more contains the queued sound cloud that should have played after.
Mixed playlist are currently a mess but are used :( 
I know from feedback that queuing music videos inside music playlist is pretty common, don't know how much use plugins vs real files.

If there's playlist switch during playback and not only at start there should be a playlist content migration too.

**Comment by enen92** [comment]
Looking at the QUEUE_NEXT_ITEM event it seems kodi doesn't even allow you to queue a video into a music playlist or queue a song into a video playlist:

https://github.com/xbmc/xbmc/blob/master/xbmc/Application.cpp#L3899-L3908

**Comment by Tolriq** [comment]
@enen92 It works via JSON it uses TMSG_PLAYLISTPLAYER_ADD
https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/PlaylistOperations.cpp#L123

-> https://github.com/xbmc/xbmc/blob/master/xbmc/PlayListPlayer.cpp#L804
-> https://github.com/xbmc/xbmc/blob/master/xbmc/PlayListPlayer.cpp#L656

The joy of Kodi internals 4 ways to do each things depending on entry point.

**Comment by enen92** [comment]
Do you mind sharing a bit more details on the test case? I am not able to queue any video from plugins into an already playing playlist (using yatse). What I am doing:

- Install a music addon (e.g. RNE podcasts)
- Install a video addon (e.g. RTP play)

Via yatse I navigate to addons -> rne podcasts -> select some folder containing items -> select a few and select queue (so audio starts playing)

Then I navigate to addons -> rtp play -> programs -> select something -> random item -> queue

Basically nothing happens, item is ignored. What am I doing wrong?

EIther way, I agree with you, maybe this can be restricted to the single case it tries to solve (first time playing a playlist, audio assumed but we got a video instead).

**Comment by Tolriq** [comment]
@enen92 long time not used that I'd need Yatse and Kodi logs to see what's happening,
Yatse already have a few tricks to workaround this issue and assign proper playlist for plugins when it can, so it's possible in that case Yatse directly queue in the wrong playlist and not the active one.

**Comment by enen92** [comment]
I think I'll close this for now and try to come up with an alternative proposal.

**Comment by Tolriq** [comment]
@enen92 I've taken time to check Yatse all workarounds and I've removed most of my mixed playlist support since quite some time, I directly queue in the proper queues for V18+. From a quick look at Kore source it seems they only do that also maybe @poisdeux can confirm they always queue in media type playlist and not activeplaylist.

That would mean that this is solution that can unlock your other issues can be an acceptable first step.

**Comment by enen92** [comment]
Either way, the original code I want to remove is correct in the current state. If we are requesting a <b>single</b> file for playback and the type is a plugin, it needs to be resolved before assuming which type it has. Since it is executed always in the mainthread, it should bring the busydialog (it's an expensive operation the user might want to cancel). So, to solve the issue I guess we need some kind of busyexecutor/queue system to make sure only one busydialog is active at any given time (plus remove the gui dependency from plugindirectory) and submit actions there when needed.

The whole playlist stuff is a complete mess in the current state as it tries to always tie items to a given type (and any playback request always goes to playlistplayer first). Looks like the same code path has grown and was reused for all kinds of stuff. Playlists are lists of items, apart from windows that might be specific to a given type (e.g. music playlists), I don't see why having a type (MUSIC/VIDEO) is even needed. Either way, the code change of this pull request only makes it worse.

**Comment by Tolriq** [comment]
Isn't this exactly the purpose of CJobQueue and https://github.com/xbmc/xbmc/blob/master/xbmc/utils/JobManager.cpp ?
